nltk/translate/ribes_score.py

```diff
@@ -0,0 +1,311 @@
+        p1 = 1.0 * len(worder) / len(hypothesis)
+        
+        _ribes = nkt * (p1 ** alpha) *  (bp ** beta)
+        
+        if _ribes > _best_ribes: # Keeps the best score.
+            _best_ribes = _ribes
+        
+    return _best_ribes, p1, bp, worder, nkt
+
+
+def position_of_ngram(ngram, sentence):
+    """
+    This function returns the position of the first instance of the ngram 
+    appearing in a sentence.
+    
+    Note that one could also use string as follows but the code is a little
+    convoluted with type casting back and forth:
+        
+        char_pos = ' '.join(sent)[:' '.join(sent).index(' '.join(ngram))]
+        word_pos = char_pos.count(' ')
+        
+    Another way to conceive this is:
+    
+        return next(i for i, ng in enumerate(ngrams(sentence, len(ngram))) 
+                    if ng == ngram)
+                    
+    :param ngram: The ngram that needs to be searched
+    :type ngram: tuple
+    :param sentence: The list of tokens to search from.
+    :type sentence: list(str)
+    """
+    # Iterates through the ngrams in sentence.
+    for i,sublist in enumerate(ngrams(sentence, len(ngram))):
+        # Returns the index of the word when ngram matches.
+        if ngram == sublist:
+            return i
+
+
+def word_rank_alignment(reference, hypothesis, character_based=False):
+    """    
+    This is the word rank alignment algorithm described in the paper to produce
+    the *worder* list, i.e. a list of word indices of the hypothesis word orders 
+    w.r.t. the list of reference words.
+    
+    Below is (H0, R0) example from the Isozaki et al. 2010 paper, 
+    note the examples are indexed from 1th but the results here starts from 0th:
+    
+        >>> ref = str('he was interested in world history because he '
+        ... 'read the book').split()
+        >>> hyp = str('he read the book because he was interested in world '
+        ... 'history').split()
+        >>> word_rank_alignment(ref, hyp)
+        [7, 8, 9, 10, 6, 0, 1, 2, 3, 4, 5]
+        
+    The (H1, R1) example from the paper, note the 0th index:
+    
+        >>> ref = 'John hit Bob yesterday'.split()
+        >>> hyp = 'Bob hit John yesterday'.split()
+        >>> word_rank_alignment(ref, hyp)
+        [2, 1, 0, 3]
+
+    Here is the (H2, R2) example from the paper, note the 0th index here too:
+    
+        >>> ref = 'the boy read the book'.split()
+        >>> hyp = 'the book was read by the boy'.split()
+        >>> word_rank_alignment(ref, hyp)
+        [3, 4, 2, 0, 1]
+        
+    :param reference: a reference sentence
+    :type reference: list(str)
+    :param hypothesis: a hypothesis sentence
+    :type hypothesis: list(str)
+    """
+    worder = []
+    hyp_len = len(hypothesis)
+    # Stores a list of possible ngrams from the reference sentence.
+    # This is used for matching context window later in the algorithm.
+    ref_ngrams = []
+    hyp_ngrams = []
+    for n in range(1, len(reference)+1):
```

Comment:
The current implementation considers ngrams of any length. The results reported in the paper considers all ngram length otherwise it will be hard to achieve the correlation results they get. Only in the examples, they want to simplify the explanation and it uses unigrams and bigrams.